### PR TITLE
Qos inter orch

### PR DESCRIPTION
**Pull request by borditamas**

### CODE REVIEW

File: qos-monitor/src/main/java/eu/arrowhead/core/qos/service/PingService.java [hunk 1]
```diff
@@ -71,4 +77,40 @@
...
+			break;
+		case JITTER_WITHOUT_TIMEOUT:
+			data.sort((final QoSIntraPingMeasurementResponseDTO m1, final QoSIntraPingMeasurementResponseDTO m2) -> m1.getJitterWithoutTimeout() - m2.getJitterWithoutTimeout());
+			break;
+		case LOST_PER_MEASUREMENT_PERCENT:
+			data.sort((final QoSIntraPingMeasurementResponseDTO m1, final QoSIntraPingMeasurementResponseDTO m2) -> m1.getLostPerMeasurementPercent() - m2.getLostPerMeasurementPercent());
+			break;
+		}
+		
+		return data.get(data.size() / 2);
```
Comment (borditamas) [inline qos-monitor/src/main/java/eu/arrowhead/core/qos/service/PingService.java:150]: Should cover the case of empty list

File: orchestrator/src/main/java/eu/arrowhead/core/qos/manager/impl/QoSManagerImpl.java [hunk 2]
```diff
@@ -158,6 +180,82 @@ public void confirmReservation(final OrchestrationResultDTO selected, final List
...
 	
+	//-------------------------------------------------------------------------------------------------
+	@Override
+	public List<GSDPollResponseDTO> preVerifyInterCloudServices(final List<GSDPollResponseDTO> gsdList, final OrchestrationFormRequestDTO request) {
+		logger.debug("preVerifyInterCloudServices started ...");
+		
+		Assert.notNull(gsdList, "'gsdList' is null.");
+		Assert.notNull(request, "'request' is null.");
+		
+		final QoSIntraPingMeasurementResponseDTO localMedianPingMeasurement = orchestratorDriver.getIntraPingMedianMeasurement(QoSMeasurementAttribute.MEAN_RESPONSE_TIME_WITHOUT_TIMEOUT);
```
Comment (borditamas) [inline orchestrator/src/main/java/eu/arrowhead/core/qos/manager/impl/QoSManagerImpl.java:191]: default from app.prop if no median available

File: gatekeeper/src/main/java/eu/arrowhead/core/gatekeeper/service/GatekeeperService.java [hunk 3]
```diff
@@ -208,10 +257,15 @@ public ICNResultDTO initICN(final ICNRequestFormDTO form) {
 		
 		final Cloud targetCloud = gatekeeperDBService.getCloudById(form.getTargetCloudId());
 		final CloudRequestDTO requesterCloud = getOwnCloud();
-		final List<RelayRequestDTO> preferredRelays = getPreferredRelays(targetCloud);
+		final List<RelayRequestDTO> preferredRelays;
+		if (form.getNegotiationFlags().get(Flag.ENABLE_QOS)) {
+			preferredRelays = form.getPreferredGatewayRelays();
```
Comment (borditamas) [inline gatekeeper/src/main/java/eu/arrowhead/core/gatekeeper/service/GatekeeperService.java]: Intersection of normal preferred and verified preferred. In no relay remain then only verified relays

File: gatekeeper/src/main/java/eu/arrowhead/core/gatekeeper/service/GatekeeperService.java [hunk 4]
```diff
@@ -280,23 +340,75 @@ public ICNProposalResponseDTO doICN(final ICNProposalRequestDTO request) {
...
+		if (needReservation) {
+			// filter out reserved providers and lock remained results temporary 
+			final QoSTemporaryLockResponseDTO lockedResults = gatekeeperDriver.sendQoSTemporaryLockRequest(new QoSTemporaryLockRequestDTO(request.getRequesterSystem(),
+																																		  orchestrationResponse.getResponse()));
+			orchestrationResponse.setResponse(lockedResults.getResponse());
+			if (orchestrationResponse.getResponse().isEmpty()) { // no usable results
+				return new ICNProposalResponseDTO();
+			}
+		} else {
+			// filter out reserved providers
```
Comment (borditamas) [inline gatekeeper/src/main/java/eu/arrowhead/core/gatekeeper/service/GatekeeperService.java]: this filter should happen right before one of the returns/matchmakings

File: gatekeeper/src/main/java/eu/arrowhead/core/gatekeeper/service/GatekeeperService.java [hunk 5]
```diff
@@ -280,23 +340,75 @@ public ICNProposalResponseDTO doICN(final ICNProposalRequestDTO request) {
...
+		}
+		
 		if (!gatewayIsMandatory) {
-			return new ICNProposalResponseDTO(orchestrationResponse.getResponse());
+			if (!needReservation) {
+				return new ICNProposalResponseDTO(orchestrationResponse.getResponse());
+			} else {
+				// If QoS enabled, then preferred systems defined in the request already contains only pre-verified systems by QoS Manager of requester cloud,
+				// therefore we have to pick and reserve one from there.
+				final OrchestrationResultDTO selectedResult = icnProviderMatchmaker.doMatchmaking(orchestrationResponse.getResponse(),
```
Comment (borditamas) [inline gatekeeper/src/main/java/eu/arrowhead/core/gatekeeper/service/GatekeeperService.java:383]: to cover null value

File: gatekeeper/src/main/java/eu/arrowhead/core/gatekeeper/service/GatekeeperService.java [hunk 6]
```diff
@@ -177,27 +193,60 @@ public GSDPollResponseDTO doGSDPoll(final GSDPollRequestDTO request) {
...
 					}
 				}
 				
-				numOfProviders++;	
+				if (!request.needQoSMeasurements()) {
+					availableInterfaces.addAll(providerInterfaces);
+					numOfProviders++;
+				} else {
+					try {
+						final QoSIntraPingMeasurementResponseDTO pingMeasurement = gatekeeperDriver.getQoSIntraPingMeasurementsForLocalSystem(providerId);
```
Comment (borditamas) [inline gatekeeper/src/main/java/eu/arrowhead/core/gatekeeper/service/GatekeeperService.java:220]: to cover empty response (skip also)

**Review by rbocsi — CHANGES_REQUESTED** [review]
Please fix those issues we talked about in the Skype meeting

File: core-common/src/main/java/eu/arrowhead/common/CoreCommonConstants.java [hunk 7]
```diff
@@ -134,6 +134,13 @@
...
 	public static final String QOS_NOT_MEASURED_SYSTEM_VERIFY_RESULT = "qos_not_measured_system_verify_result"; // result of verify if a system has no measurement records
 	public static final String $QOS_NOT_MEASURED_SYSTEM_VERIFY_RESULT = "${" + QOS_NOT_MEASURED_SYSTEM_VERIFY_RESULT + ":" + CoreDefaults.DEFAULT_QOS_NOT_MEASURED_SYSTEM_VERIFY_RESULT + "}";
+	public static final String $QOS_DEFAULT_REFERENCE_MIN_RESPONSE_TIME_WD = "${default.min_response_time:" + CoreDefaults.DEFAULT_QOS_DEFAULT_REFERENCE_MIN_RESPONSE_TIME + "}";
+	public static final String $QOS_DEFAULT_REFERENCE_MAX_RESPONSE_TIME_WD = "${default.max_response_time:" + CoreDefaults.DEFAULT_QOS_DEFAULT_REFERENCE_MAX_RESPONSE_TIME + "}";
+	public static final String $QOS_DEFAULT_REFERENCE_MEAN_RESPONSE_TIME_WITH_TIMEOUT_WD = "${default.mean_response_time_with_timeout:" + CoreDefaults.DEFAULT_QOS_DEFAULT_REFERENCE_MEAN_RESPONSE_TIME_WITH_TIMEOUT + "}";
+	public static final String $QOS_DEFAULT_REFERENCE_MEAN_RESPONSE_TIME_WITHOUT_TIMEOUT_WD = "${default.mean_response_time_without_timeout:" + CoreDefaults.DEFAULT_QOS_DEFAULT_REFERENCE_MEAN_RESPONSE_TIME_WITHOUT_TIMEOUT + "}";
```
Comment (rbocsi) [inline core-common/src/main/java/eu/arrowhead/common/CoreCommonConstants.java]: I think these two should be the same

File: core-common/src/main/java/eu/arrowhead/common/CoreCommonConstants.java [hunk 8]
```diff
@@ -134,6 +134,13 @@
...
+	public static final String $QOS_DEFAULT_REFERENCE_JITTER_WITH_TIMEOUT_WD = "${default.jitter_with_timeout:" + CoreDefaults.DEFAULT_QOS_DEFAULT_REFERENCE_JITTER_WITH_TIMEOUT + "}";
+	public static final String $QOS_DEFAULT_REFERENCE_JITTER_WITHOUT_TIMEOUT_WD = "${default.jitter_without_timeout:" + CoreDefaults.DEFAULT_QOS_DEFAULT_REFERENCE_JITTER_WITHOUT_TIMEOUT + "}";
```
Comment (rbocsi) [inline core-common/src/main/java/eu/arrowhead/common/CoreCommonConstants.java]: same here

File: core-common/src/main/java/eu/arrowhead/common/CoreCommonConstants.java [hunk 9]
```diff
@@ -134,6 +134,13 @@
...
+	public static final String $QOS_DEFAULT_REFERENCE_LOST_PET_MEASUREMENT_PERCENT_WD = "${default.lost_per_measurement_percent:" + CoreDefaults.DEFAULT_QOS_DEFAULT_REFERENCE_LOST_PET_MEASUREMENT_PERCENT + "}";
```
Comment (rbocsi) [inline core-common/src/main/java/eu/arrowhead/common/CoreCommonConstants.java]: typo in constant name (PET)

File: core-common/src/main/java/eu/arrowhead/common/CoreDefaults.java [hunk 10]
```diff
@@ -56,6 +56,13 @@
 	public static final int DEFAULT_QOS_MAX_RESERVATION_DURATION = 3600; // in seconds
 	public static final int DEFAULT_QOS_PING_MEASUREMENT_CACHE_THRESHOLD = 600; // in seconds
 	public static final boolean DEFAULT_QOS_NOT_MEASURED_SYSTEM_VERIFY_RESULT = true;
+	public static final int DEFAULT_QOS_DEFAULT_REFERENCE_MIN_RESPONSE_TIME = 30;
+	public static final int DEFAULT_QOS_DEFAULT_REFERENCE_MAX_RESPONSE_TIME = 34;
+	public static final int DEFAULT_QOS_DEFAULT_REFERENCE_MEAN_RESPONSE_TIME_WITH_TIMEOUT = 32;
+	public static final int DEFAULT_QOS_DEFAULT_REFERENCE_MEAN_RESPONSE_TIME_WITHOUT_TIMEOUT = 32;
+	public static final int DEFAULT_QOS_DEFAULT_REFERENCE_JITTER_WITH_TIMEOUT = 2;
+	public static final int DEFAULT_QOS_DEFAULT_REFERENCE_JITTER_WITHOUT_TIMEOUT = 2;
+	public static final int DEFAULT_QOS_DEFAULT_REFERENCE_LOST_PET_MEASUREMENT_PERCENT = 0;
```
Comment (rbocsi) [inline core-common/src/main/java/eu/arrowhead/common/CoreDefaults.java]: typo here PET

**Review by rbocsi — APPROVED** [review]
There are some minor issues (see inline comments)
Also, no typo fix in QosMonitorController's method (getMeadian or something)

**Pull request merged**